File: packages/cli/src/report.ts
// Aggregation — and the privacy boundary.
//
// Message text enters this module and does not leave it. Everything below is a
// count, a rate, or a matched lexicon word; nothing that reaches the browser,
// the cache, or stdout carries what was actually written.
//
// The user's side and the agent's side are aggregated separately. `totals`,
// `daily`, `heatmap`, and `projects` are all user-scoped; the agent gets its
// own counters rather than being folded into the headline number.

import { existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type {
  AgentDayStat,
  AgentHarnessStats,
  AgentTotals,
  DayStat,
  HeatCell,
  HarnessStats,
  ProjectStat,
  Report,
  Tier,
  Totals,
  WordStat,
} from "@salt/core";
import type { Matcher } from "./match.js";
import { cmpCodePoints, splitWhitespace } from "./text.js";
import type { Message, ScanStats } from "./types.js";
import { ALL_HARNESSES } from "./types.js";
import { VERSION } from "./version.js";

interface Bucket {
  prompts: number;
  swears: number;
  prompts_with_swear: number;
}

function bucketAdd(map: Map<string, Bucket>, key: string, swears: number): void {
  let b = map.get(key);
  if (!b) {
    b = { prompts: 0, swears: 0, prompts_with_swear: 0 };
    map.set(key, b);
  }
  b.prompts += 1;
  b.swears += swears;
  if (swears > 0) b.prompts_with_swear += 1;
}

function pairAdd(map: Map<string, [number, number]>, key: string, swears: number): void {
  const e = map.get(key) ?? [0, 0];
  e[0] += 1;
  e[1] += swears;
  map.set(key, e);
}

function rate(swears: number, prompts: number): number {
  return prompts === 0 ? 0 : (100 * swears) / prompts;
}

function wordCount(text: string): number {
  return splitWhitespace(text).length;
}

/**
 * Ranks matched words by frequency, ties broken alphabetically so the order is
 * stable across runs.
 */
function rankWords(counts: Map<string, { tier: Tier; count: number }>, total: number): WordStat[] {
  const out: WordStat[] = [...counts.entries()].map(([word, { tier, count }]) => ({
    word,
    tier,
    count,
    share: total === 0 ? 0 : count / total,
  }));
  out.sort((a, b) => b.count - a.count || cmpCodePoints(a.word, b.word));
  return out;
}

function countHits(
  counts: Map<string, { tier: Tier; count: number }>,
  hits: { word: string; tier: Tier }[],
): void {
  for (const h of hits) {
    const e = counts.get(h.word);
    if (e) e.count += 1;
    else counts.set(h.word, { tier: h.tier, count: 1 });
  }
}

export function buildReport(messages: Message[], stats: ScanStats, matcher: Matcher): Report {
  const totals: Totals = {
    sessions: 0,
    prompts: 0,
    words: 0,
    swears: 0,
    prompts_with_swear: 0,
    swears_per_100_prompts: 0,
  };

  const byHarness = new Map<string, Bucket>();
  const byDay = new Map<string, [number, number]>();
  const byCell = new Map<string, [number, number]>();
  const byProject = new Map<string, [number, number]>();
  const wordCounts = new Map<string, { tier: Tier; count: number }>();
  const sessions = new Set<string>();
  let sessionPrecision = 0;

  const agent: AgentTotals = {
    messages: 0,
    words: 0,
    swears: 0,
    messages_with_swear: 0,
    swears_per_100_messages: 0,
  };
  const agentByHarness = new Map<string, Bucket>();
  const agentByDay = new Map<string, [number, number]>();
  const agentWordCounts = new Map<string, { tier: Tier; count: number }>();

  for (const p of messages) {
    const hits = matcher.find(p.text);
    const swears = hits.length;
    const date = new Date(p.ts).toISOString().slice(0, 10);

    if (p.role === "agent") {
      agent.messages += 1;
      agent.words += wordCount(p.text);
      agent.swears += swears;
      if (swears > 0) agent.messages_with_swear += 1;
      countHits(agentWordCounts, hits);
      bucketAdd(agentByHarness, p.harness, swears);
      pairAdd(agentByDay, `${date}\x00${p.harness}`, swears);
      continue;
    }

    totals.prompts += 1;
    totals.words += wordCount(p.text);
    totals.swears += swears;
    if (swears > 0) totals.prompts_with_swear += 1;

    countHits(wordCounts, hits);

    sessions.add(`${p.harness}\x00${p.sessionId}`);
    bucketAdd(byHarness, p.harness, swears);

    // `daily` buckets by UTC date while the heatmap uses local wall-clock time
    // — a deliberate inconsistency inherited from v1: the timeline is a shared
    // calendar, the heatmap is about the shape of this user's day.
    pairAdd(byDay, date, swears);

    // Cursor stores no per-message time, so those prompts would smear the
    // heatmap across whatever hour the chat was last touched.
    if (p.tsPrecision === "exact") {
      const local = new Date(p.ts);
      pairAdd(byCell, `${(local.getDay() + 6) % 7}\x00${local.getHours()}`, swears);
    } else {
      sessionPrecision += 1;
    }

    if (p.cwd !== undefined) {
      const name = projectName(p.cwd);
      if (name !== null) pairAdd(byProject, name, swears);
    }
  }

  totals.sessions = sessions.size;
  totals.swears_per_100_prompts = rate(totals.swears, totals.prompts);
  agent.swears_per_100_messages = rate(agent.swears, agent.messages);

  const projects: ProjectStat[] = [...byProject.entries()].map(([name, [prompts, swears]]) => ({
    name,
    prompts,
    swears,
    rate: rate(swears, prompts),
  }));
  projects.sort(
    (a, b) => b.swears - a.swears || b.prompts - a.prompts || cmpCodePoints(a.name, b.name),
  );

  const heatmap: HeatCell[] = [...byCell.entries()].map(([key, [prompts, swears]]) => {
    const [dow, hour] = key.split("\x00").map(Number);
    return { dow, hour, prompts, swears };
  });
  heatmap.sort((a, b) => a.dow - b.dow || a.hour - b.hour);

  const harnessStats = (map: Map<string, Bucket>): HarnessStats[] =>
    ALL_HARNESSES.filter((h) => map.has(h)).map((h) => {
      const b = map.get(h)!;
      return {
        harness: h,
        prompts: b.prompts,
        swears: b.swears,
        prompts_with_swear: b.prompts_with_swear,
        rate: rate(b.swears, b.prompts),
      };
    });

  const agentByHarnessOut: AgentHarnessStats[] = harnessStats(agentByHarness).map((s) => ({
    harness: s.harness,
    messages: s.prompts,
    swears: s.swears,
    messages_with_swear: s.prompts_with_swear,
    rate: s.rate,
  }));

  const daily: DayStat[] = [...byDay.keys()].sort().map((date) => {
    const [prompts, swears] = byDay.get(date)!;
    return { date, prompts, swears };
  });

  // Days a harness was idle are absent rather than zero-filled; the chart
  // reads a gap as zero.
  const agentDaily: AgentDayStat[] = [...agentByDay.keys()].sort().map((key) => {
    const [date, harness] = key.split("\x00");
    const [messages, swears] = agentByDay.get(key)!;
    return { date, harness: harness as AgentDayStat["harness"], messages, swears };
  });

  const notes: string[] = [];
  if (sessionPrecision > 0) {
    notes.push(
      `${sessionPrecision} Cursor prompts have session-level timestamps only and are excluded from the heatmap.`,
    );
  }
  if (stats.files_failed > 0) {
    notes.push(`${stats.files_failed} files could not be read.`);
  }
  if (stats.duplicates_dropped > 0) {
    notes.push(
      `${stats.duplicates_dropped} replayed prompts were collapsed; agent harnesses rewrite session history on fork and resume.`,
    );
  }

  // Field order mirrors the serialised Rust struct, byte for byte.
  return {
    generated_at: new Date().toISOString(),
    version: VERSION,
    totals,
    by_harness: harnessStats(byHarness),
    top_words: rankWords(wordCounts, totals.swears),
    daily,
    heatmap,
    projects,
    agent,
    agent_by_harness: agentByHarnessOut,
    agent_daily: agentDaily,
    agent_top_words: rankWords(agentWordCounts, agent.swears),
    coverage: {
      files_scanned: stats.files_scanned,
      files_failed: stats.files_failed,
      bytes_scanned: stats.bytes_scanned,
      duplicates_dropped: stats.duplicates_dropped,
      session_precision_prompts: sessionPrecision,
      notes,
    },
  };
}

/**
 * Reduces a working directory to a shareable project name: the basename of the
 * enclosing git repository, else the leaf directory.
 *
 * Paths compare the way v1's `std::path::Path` did — `//` collapsed, a
 * non-leading `.` normalized away, `..` left alone — so no spelling of the
 * home directory (`/Users/name/.`, `/Users//name`) slips past the exclusion.
 */
export function projectName(cwd: string): string | null {
  const path = components(cwd);
  // The home directory is not a project; prompts typed there would otherwise
  // rank as one and expose the account name.
  const home = components(homedir());
  if (path.root === home.root && listEq(path.parts, home.parts)) return null;

  // Walk to the filesystem root — or, for a relative path, to "" — exactly
  // the ancestor chain `Path::parent` produces.
  for (let n = path.parts.length; n >= 0; n--) {
    if (existsSync(join(pathOf(path.root, path.parts.slice(0, n)), ".git"))) {
      return fileName(path.parts.slice(0, n));
    }
  }
  return fileName(path.parts);
}

/** Normalized path components plus whether the path was absolute. */
function components(path: string): { root: boolean; parts: string[] } {
  const root = path.startsWith("/");
  const segs = path.split("/").filter((seg) => seg !== "");
  // `.` survives only as the first component of a relative path.
  const parts = segs.filter((seg, i) => seg !== "." || (i === 0 && !root));
  return { root, parts };
}

function pathOf(root: boolean, parts: string[]): string {
  return (root ? "/" : "") + parts.join("/");
}

/** `Path::file_name`: none for an empty path or one ending in `.` or `..`. */
function fileName(parts: string[]): string | null {
  const last = parts[parts.length - 1];
  return last === undefined || last === "." || last === ".." ? null : last;
}

function listEq(a: string[], b: string[]): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}

File: packages/cli/src/report.rs
//! Aggregation — and the privacy boundary.
//!
//! Message text enters this module and does not leave it. Everything below is a
//! count, a rate, or a matched lexicon word; nothing that reaches the browser,
//! the cache, or stdout carries what was actually written. `build` takes the
//! messages by value so the prompt text is dropped on return rather than left
//! alive in the caller's vector.
//!
//! The user's side and the agent's side are aggregated separately. `totals`,
//! `daily`, and `projects` are all user-scoped; the agent gets its own counters
//! rather than being folded into the headline number.

use std::collections::{BTreeMap, HashMap, HashSet};
use std::ffi::OsStr;
use std::path::Path;

use chrono::{DateTime, Local, Utc};

use crate::matcher::{Hit, Matcher};
use crate::types::{
    AgentDayStat, AgentHarnessStats, AgentTotals, Coverage, DayStat, Harness, HarnessStats,
    Message, ProjectStat, Report, Role, ScanStats, Tier, Totals, TsPrecision, WordStat,
    ALL_HARNESSES,
};

#[derive(Default, Clone, Copy)]
struct Bucket {
    prompts: u64,
    swears: u64,
    prompts_with_swear: u64,
}

impl Bucket {
    fn add(&mut self, swears: u64) {
        self.prompts += 1;
        self.swears += swears;
        if swears > 0 {
            self.prompts_with_swear += 1;
        }
    }
}

/// Days carry a third counter: prompts, swears, and the severity weight of
/// those swears, which is what the calendar shades by.
#[derive(Default, Clone, Copy)]
struct Day {
    prompts: u64,
    swears: u64,
    weight: u64,
}

fn rate(swears: u64, prompts: u64) -> f64 {
    if prompts == 0 {
        0.0
    } else {
        // Multiply before dividing, as the TypeScript does, so the rounding is
        // bit-identical between the two implementations.
        100.0 * swears as f64 / prompts as f64
    }
}

/// The publisher's own calendar day. A prompt typed at 11pm belongs to that
/// evening, not to the next UTC date, and this is a log of one person's days.
fn local_date(ts: i64) -> String {
    match DateTime::from_timestamp_millis(ts) {
        Some(t) => t.with_timezone(&Local).format("%Y-%m-%d").to_string(),
        // Beyond chrono's range; JS would produce "NaN-NaN-NaN" here. Keeping a
        // parseable string means one absurd timestamp cannot break the calendar.
        None => "0000-00-00".to_string(),
    }
}

/// Ranks matched words by frequency, ties broken alphabetically so the order is
/// stable across runs. Byte order on `str` is the code point order the
/// TypeScript had to emulate.
fn rank_words(counts: HashMap<String, (Tier, u64)>, total: u64) -> Vec<WordStat> {
    let mut out: Vec<WordStat> = counts
        .into_iter()
        .map(|(word, (tier, count))| WordStat {
            word,
            tier,
            count,
            share: if total == 0 {
                0.0
            } else {
                count as f64 / total as f64
            },
        })
        .collect();
    out.sort_by(|a, b| b.count.cmp(&a.count).then_with(|| a.word.cmp(&b.word)));
    out
}

fn harness_stats(map: &HashMap<Harness, Bucket>) -> Vec<HarnessStats> {
    ALL_HARNESSES
        .iter()
        .filter_map(|h| {
            let b = map.get(h)?;
            Some(HarnessStats {
                harness: *h,
                prompts: b.prompts,
                swears: b.swears,
                prompts_with_swear: b.prompts_with_swear,
                rate: rate(b.swears, b.prompts),
            })
        })
        .collect()
}

pub fn build(messages: Vec<Message>, stats: ScanStats, matcher: &Matcher, version: &str) -> Report {
    let mut totals = Totals::default();
    let mut by_harness: HashMap<Harness, Bucket> = HashMap::new();
    let mut by_day: BTreeMap<String, Day> = BTreeMap::new();
    let mut by_project: HashMap<String, (u64, u64)> = HashMap::new();
    let mut word_counts: HashMap<String, (Tier, u64)> = HashMap::new();
    let mut sessions: HashSet<(Harness, String)> = HashSet::new();
    let mut session_precision: u64 = 0;

    let mut agent = AgentTotals::default();
    let mut agent_by_harness_map: HashMap<Harness, Bucket> = HashMap::new();
    let mut agent_by_day: BTreeMap<(String, Harness), (u64, u64)> = BTreeMap::new();
    let mut agent_word_counts: HashMap<String, (Tier, u64)> = HashMap::new();

    // One repo lookup per distinct cwd. `project_name` walks the ancestor chain
    // stat'ing `.git`, and a busy project turns up in thousands of prompts; v1
    // repeated the whole chain for every one of them.
    let mut project_cache: HashMap<String, Option<String>> = HashMap::new();
    let home = dirs::home_dir();

    for p in messages {
        let hits: Vec<Hit> = matcher.find(&p.text);
        let swears = hits.len() as u64;
        let date = local_date(p.ts);

        if p.role == Role::Agent {
            agent.messages += 1;
            agent.words += p.text.split_whitespace().count() as u64;
            agent.swears += swears;
            if swears > 0 {
                agent.messages_with_swear += 1;
            }
            Matcher::tally(&hits, &mut agent_word_counts);
            agent_by_harness_map
                .entry(p.harness)
                .or_default()
                .add(swears);
            let e = agent_by_day.entry((date, p.harness)).or_default();
            e.0 += 1;
            e.1 += swears;
            continue;
        }

        totals.prompts += 1;
        totals.words += p.text.split_whitespace().count() as u64;
        totals.swears += swears;
        if swears > 0 {
            totals.prompts_with_swear += 1;
        }

        Matcher::tally(&hits, &mut word_counts);

        sessions.insert((p.harness, p.session_id));
        by_harness.entry(p.harness).or_default().add(swears);

        let day = by_day.entry(date).or_default();
        day.prompts += 1;
        day.swears += swears;
        for h in &hits {
            day.weight += h.tier.weight();
        }

        // Cursor stores no per-message time, only when the chat was last touched.
        // At day resolution that is close enough to keep — the calendar counts it,
        // and coverage says so.
        if p.ts_precision != TsPrecision::Exact {
            session_precision += 1;
        }

        if let Some(cwd) = p.cwd {
            let name = project_cache
                .entry(cwd)
                .or_insert_with_key(|cwd| project_name_with_home(cwd, home.as_deref()))
                .clone();
            if let Some(name) = name {
                let e = by_project.entry(name).or_insert((0, 0));
                e.0 += 1;
                e.1 += swears;
            }
        }
    }

    totals.sessions = sessions.len() as u64;
    totals.swears_per_100_prompts = rate(totals.swears, totals.prompts);
    agent.swears_per_100_messages = rate(agent.swears, agent.messages);

    let mut projects: Vec<ProjectStat> = by_project
        .into_iter()
        .map(|(name, (prompts, swears))| ProjectStat {
            name,
            prompts,
            swears,
            rate: rate(swears, prompts),
        })
        .collect();
    projects.sort_by(|a, b| {
        b.swears
            .cmp(&a.swears)
            .then_with(|| b.prompts.cmp(&a.prompts))
            .then_with(|| a.name.cmp(&b.name))
    });

    let agent_by_harness: Vec<AgentHarnessStats> = harness_stats(&agent_by_harness_map)
        .into_iter()
        .map(|s| AgentHarnessStats {
            harness: s.harness,
            messages: s.prompts,
            swears: s.swears,
            messages_with_swear: s.prompts_with_swear,
            rate: s.rate,
        })
        .collect();

    let daily: Vec<DayStat> = by_day
        .into_iter()
        .map(|(date, d)| DayStat {
            date,
            prompts: d.prompts,
            swears: d.swears,
            weight: d.weight,
        })
        .collect();

    // Days a harness was idle are absent rather than zero-filled; the chart
    // reads a gap as zero.
    let agent_daily: Vec<AgentDayStat> = agent_by_day
        .into_iter()
        .map(|((date, harness), (messages, swears))| AgentDayStat {
            date,
            harness,
            messages,
            swears,
        })
        .collect();

    let mut notes: Vec<String> = Vec::new();
    if session_precision > 0 {
        notes.push(format!(
            "{session_precision} Cursor prompts have session-level timestamps only; they are dated by when their chat was last touched."
        ));
    }
    if stats.files_failed > 0 {
        // Not pluralised, deliberately: the TypeScript wording is the wire
        // contract, so "1 files could not be read." is the correct output.
        notes.push(format!("{} files could not be read.", stats.files_failed));
    }
    if stats.duplicates_dropped > 0 {
        notes.push(format!(
            "{} replayed prompts were collapsed; agent harnesses rewrite session history on fork and resume.",
            stats.duplicates_dropped
        ));
    }

    Report {
        // JS `Date#toISOString`: always exactly three fractional digits.
        generated_at: Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string(),
        version: version.to_string(),
        by_harness: harness_stats(&by_harness),
        top_words: rank_words(word_counts, totals.swears),
        totals,
        daily,
        projects,
        agent_by_harness,
        agent_daily,
        agent_top_words: rank_words(agent_word_counts, agent.swears),
        agent,
        coverage: Coverage {
            files_scanned: stats.files_scanned,
            files_failed: stats.files_failed,
            bytes_scanned: stats.bytes_scanned,
            duplicates_dropped: stats.duplicates_dropped,
            session_precision_prompts: session_precision,
            notes,
        },
    }
}

/// Reduces a working directory to a shareable project name: the basename of the
/// enclosing git repository, else the leaf directory.
///
/// This used to normalise path components by hand so that no spelling of the
/// home directory (`/Users/name/.`, `/Users//name`) slipped past the exclusion.
/// `Path` already does exactly that — it compares by components, so repeated
/// separators and non-leading `.` are equal by construction — and its
/// `ancestors()` is the same chain the hand-rolled walk produced.
pub fn project_name(cwd: &str) -> Option<String> {
    project_name_with_home(cwd, dirs::home_dir().as_deref())
}

/// Same, with the home directory resolved once by the caller — `build` looks it
/// up per report rather than per message.
fn project_name_with_home(cwd: &str, home: Option<&Path>) -> Option<String> {
    let path = Path::new(cwd);
    // The home directory is not a project; prompts typed there would otherwise
    // rank as one and expose the account name.
    if home == Some(path) {
        return None;
    }

    // A `.git` *file* counts too: that is what a worktree or submodule
    // checkout has.
    path.ancestors()
        .find(|dir| dir.join(".git").exists())
        .unwrap_or(path)
        .file_name()
        .and_then(OsStr::to_str)
        .map(str::to_owned)
}
